File: docs/aimdk/src/py_examples/py_examples/get_map.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.executors import SingleThreadedExecutor
from aimdk_msgs.srv import GetStoredMapByName
from std_msgs.msg import Header


class MapClient(Node):
    def __init__(self):
        super().__init__('get_map_client')
        self.client = self.create_client(
            GetStoredMapByName,
            '/aimdk_5Fmsgs/srv/GetStoredMapByName'
        )

    def send_request(self, map_name):
        while not self.client.wait_for_service(timeout_sec=1.0):
            if not rclpy.ok():
                self.get_logger().error("Interrupted while waiting for service.")
                return
            self.get_logger().info("Service not available, waiting again...")

        request = GetStoredMapByName.Request()

        request.header = Header()
        request.header.stamp = self.get_clock().now().to_msg()
        request.header.frame_id = ""

        request.map_name = map_name

        future = self.client.call_async(request)

        executor = SingleThreadedExecutor()
        executor.add_node(self)
        executor.spin_until_future_complete(future, timeout_sec=60.0)
        executor.remove_node(self)

        if future.result() is None:
            self.get_logger().error("Service call failed or timed out.")
            return

        result = future.result()

        if result.code == 0:
            self.get_logger().info("Service call succeeded")
            self.get_logger().info(f"Map path: {result.map_path}")
            self.get_logger().info(
                f"Map resolution: {result.map_info.resolution}")
            self.get_logger().info(f"Map width: {result.map_info.width}")
            self.get_logger().info(f"Map height: {result.map_info.height}")
            self.get_logger().info(f"Map id: {result.map_id}")
        else:
            self.get_logger().error(
                f"Service call failed with code: {result.code}")


def main(args=None):
    rclpy.init(args=args)

    client = MapClient()
    try:
        client.send_request("map_name")
    except KeyboardInterrupt:
        pass
    finally:
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: docs/aimdk/src/examples/src/autonomous/get_map.cpp
//
// Created by agiuser on 2026/3/2.
//

#include "aimdk_msgs/srv/get_stored_map_by_name.hpp"
#include "rclcpp/rclcpp.hpp"
#include <memory>

class MapClient : public rclcpp::Node {
public:
  MapClient() : Node("get_map_client") {
    client_ = this->create_client<aimdk_msgs::srv::GetStoredMapByName>(
        "/aimdk_5Fmsgs/srv/GetStoredMapByName");
  }

  void send_request(const std::string &map_name) {
    while (!client_->wait_for_service(std::chrono::seconds(1))) {
      if (!rclcpp::ok()) {
        RCLCPP_ERROR(this->get_logger(),
                     "Interrupted while waiting for service.");
        return;
      }
      RCLCPP_INFO(this->get_logger(),
                  "Service not available, waiting again...");
    }

    auto request =
        std::make_shared<aimdk_msgs::srv::GetStoredMapByName::Request>();

    request->header.stamp = this->now();
    request->header.frame_id = "";

    request->map_name = map_name;

    auto result_future = client_->async_send_request(request);

    // Use a fresh executor to spin, avoiding any pre-existing state.
    rclcpp::executors::SingleThreadedExecutor exec;
    exec.add_node(shared_from_this());
    if (exec.spin_until_future_complete(result_future,
                                        std::chrono::seconds(10)) !=
        rclcpp::FutureReturnCode::SUCCESS) {
      RCLCPP_ERROR(this->get_logger(), "Service call failed or timed out.");
      return;
    }

    auto result = result_future.get();

    if (result->code == 0) {
      RCLCPP_INFO(this->get_logger(), "Service call succeeded");
      RCLCPP_INFO(this->get_logger(), "Map path: %s", result->map_path.c_str());
      RCLCPP_INFO(this->get_logger(), "Map resolution: %f",
                  result->map_info.resolution);
      RCLCPP_INFO(this->get_logger(), "Map width: %u", result->map_info.width);
      RCLCPP_INFO(this->get_logger(), "Map height: %u",
                  result->map_info.height);
      RCLCPP_INFO(this->get_logger(), "Map id: %ld", result->map_id);
    } else {
      RCLCPP_ERROR(this->get_logger(), "Service call failed with code: %lu",
                   result->code);
    }
  }

private:
  rclcpp::Client<aimdk_msgs::srv::GetStoredMapByName>::SharedPtr client_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);

  auto client = std::make_shared<MapClient>();
  client->send_request("map_name");

  rclcpp::shutdown();
  return 0;
}
